cmd/access-control/access-control.go

```diff
@@ -0,0 +1,234 @@
+package accesscontrol
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang-jwt/jwt/v4"
+	glog "github.com/magicsong/color-glog"
+	"github.com/pquerna/cachecontrol/cacheobject"
+)
+
+type PlaybackAccessControl struct {
+	gateURL string
+	cache   map[string]map[string]*PlaybackAccessControlEntry
+	*http.Client
+}
+
+type PlaybackAccessControlEntry struct {
+	Stale  time.Time
+	MaxAge time.Time
+	Allow  bool
+	mutex  sync.Mutex
+}
+
+type PlaybackAccessControlRequest struct {
+	Type   string `json:"type"`
+	Pub    string `json:"pub"`
+	Stream string `json:"stream"`
+}
+
+const UserNewTrigger = "USER_NEW"
+
+func TriggerHandler(gateURL string) http.Handler {
+	playbackAccessControl := PlaybackAccessControl{
+		gateURL,
+		make(map[string]map[string]*PlaybackAccessControlEntry),
+		&http.Client{},
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		payload, err := io.ReadAll(r.Body)
+
+		if err != nil {
+			w.Write([]byte("false"))
+			glog.Errorf("Unable to parse trigger body %v", err)
+			return
+		}
+
+		triggerName := r.Header.Get("X-Trigger")
+
+		switch triggerName {
+		case UserNewTrigger:
+			w.Write(handleUserNew(&playbackAccessControl, payload))
+			return
+		default:
+			w.Write([]byte("false"))
+			glog.Errorf("Trigger not handled %v", triggerName)
+			return
+		}
+
+	})
+}
+
+func handleUserNew(ac *PlaybackAccessControl, payload []byte) []byte {
+	lines := strings.Split(string(payload), "\n")
+
+	if len(lines) != 6 {
+		glog.Errorf("Malformed trigger payload")
+		return []byte("false")
+	}
+
+	requestURL, err := url.Parse(lines[4])
+	if err != nil {
+		glog.Errorf("Unable to parse URL %v", err)
+		return []byte("false")
+	}
+
+	playbackID := lines[0]
+	jwtToken := requestURL.Query().Get("jwt")
+
+	var pubKey string
+	if jwtToken != "" {
+		claims, err := decodeJwt(jwtToken)
+		if err != nil {
+			glog.Errorf("Unable to decode JWT token %v", err)
+			return []byte("false")
+		}
+
+		if playbackID != claims.Subject {
+			glog.Errorf("PlaybackId mismatch")
+			return []byte("false")
+		}
+
+		pubKey = claims.PublicKey
+	}
+
+	playbackAccessControl, err := getPlaybackAccessControlInfo(ac, playbackID, pubKey)
+	if err != nil {
+		glog.Errorf("Unable to get playback access control info %v", err)
+		return []byte("false")
+	}
+
+	if playbackAccessControl[pubKey].Allow {
+		return []byte("true")
+	}
+
+	return []byte("false")
+}
+
+func getPlaybackAccessControlInfo(ac *PlaybackAccessControl, playbackID, pubKey string) (map[string]*PlaybackAccessControlEntry, error) {
+	if ac.cache[playbackID] == nil ||
+		ac.cache[playbackID][pubKey] == nil ||
+		time.Now().After(ac.cache[playbackID][pubKey].Stale) {
+
+		err := cachePlaybackAccessControlInfo(ac, playbackID, pubKey)
+		if err != nil {
+			return nil, err
+		}
+
+	} else if time.Now().After(ac.cache[playbackID][pubKey].MaxAge) {
```

Comment:
@hjpotter92, @victorges would love to have your feedback on this part. ctx was added to finish tests properly and lock to avoid concurrent fetches on a single resource. We should also add a `waitGroup` to globally throttle concurrency. WDYT?

Reply:
Hey! I believe the problem you were having in the tests was something else, because the use of the context here is a noop. It is not passed anywhere, and `Done` is a read only function on the context and so it's not having any effect here AFAICT. What problem were you having that this fixed?

Regarding the mutex, there is still several places that access the `cache` map (and the maps inside it) without any locking mechanism. Looking quickly at the structs here, I believe that what you need to do is move the mutex to the root level of the struct, and make sure you grab an `RLock` before you read anything from it, and a (W) `Lock` from the cache function, just when you are about to modify it.

You should not worry about limiting concurrent requests on this PR. We'd need a huge amount of viewers until this becomes a problem or would be worth optimizing. The HTTP client will already limit the number of open connections to the same host, and so that will already provide a hard limit on concurrent request anyway. Anything beyond that would likely be over-engineering right now, and we want to get this through quickly so we should avoid that.

Sorry for my original comment that pointed that this would be a problem on the current version. I have changed my mind and do not think it is a problem worth fixing right now.